onnx/backend/test/cpp/driver/test_driver.cc

```diff
@@ -0,0 +1,180 @@
+			if (errno != ERRNO_DIR_END){
+				fprintf(stderr, "Error: cannot read directory %s when loading test data: %s\n",
+					target_dir.c_str(), strerror(errno));
+				return -1;
+			} else{
+				break;
+			}
+		}
+                std::string entry_dname = entry->d_name;
+                entry_dname = _target_dir + '/' + entry_dname + "/";
+                fetchSingleTestCase(entry_dname);
+        }
+cleanup:
+	if (directory != NULL){
+		if (closedir(directory) != 0){
+			fprintf(stderr, "Warning: failed to close directory %s when loading test data: %s\n",
+				target_dir.c_str(), strerror(errno));
+			return -1;
+		}
+	}
+	return 0;
+}
+
+std::vector<TestCase> getTestCase(const std::string& location){
+	TestDriver test_driver;
+	test_driver.fetchAllTestCases(location);
+	return test_driver.testcases_;
+}
+
+void loadSingleFile(const std::string& filename, std::string& filedata){
+	std::ifstream f(filename, std::ifstream::in);
+	if (f.is_open()){
+			std::string line;
+		getline(f, filedata);
+		while (getline(f, line)){
+			filedata += "\n" + line;
+		}
+	}
+	f.close();
+}
+
+ProtoTestCase loadSingleTestCase(const TestCase& t){
+
+	ProtoTestCase st;
+	loadSingleFile(t.model_filename_, st.raw_model_);
+	ParseProtoFromBytes(&st.model_, st.raw_model_.c_str(), st.raw_model_.size());
+
+	for (auto test_data : t.test_data_){
+		ProtoTestData proto_test_data;
+
+		for (auto input_file : test_data.input_filenames_){
+			std::string input_data = "";
+			loadSingleFile(input_file, input_data);
+			proto_test_data.raw_inputs_.push_back(input_data);
+			onnx::TensorProto input_proto;
+			ParseProtoFromBytes(&input_proto, input_data.c_str(), input_data.size());
+			proto_test_data.inputs_.push_back(input_proto);
+		}
+		for (auto output_file : test_data.output_filenames_){
+			std::string output_data = "";
+			loadSingleFile(output_file, output_data);
+			proto_test_data.raw_outputs_.push_back(output_data);
+			onnx::TensorProto output_proto;
+			ParseProtoFromBytes(&output_proto, output_data.c_str(), output_data.size());
+			proto_test_data.outputs_.push_back(output_proto);
+		}
+	}
+	return st;
+}
+
+std::vector<ProtoTestCase> loadAllTestCases(const std::string& location){
+	std::vector<TestCase> t = getTestCase(location);
+	return loadAllTestCases(t);
+}
+
+std::vector<ProtoTestCase> loadAllTestCases(const std::vector<TestCase>& t){
+	std::vector<ProtoTestCase> st;
+	for (auto i : t){
+		st.push_back(loadSingleTestCase(i));
+	}
+	return st;
+}
+
+//Need to be migrated to util in the future
+onnxTensorDescriptorV1 ProtoToOnnxTensorDescriptor(const onnx::TensorProto& proto_tensor){
+	onnxTensorDescriptorV1 onnx_tensor;
+	onnx_tensor.tag = ONNXIFI_TAG_TENSOR_DESCRIPTOR_V1;
+	onnx_tensor.name = proto_tensor.name().c_str();
+	onnx_tensor.dataType = proto_tensor.data_type();
+	onnx_tensor.memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+	onnx_tensor.dimensions = proto_tensor.dims().size();
+	onnx_tensor.shape = (uint64_t *)proto_tensor.dims().data();
```

Comment:
what is the data type of `proto_tensor.dims().data()`? this doesn't look safe

Reply:
It is a long long *, so the only difference is the "unsigned". I think this is safe?

Reply:
Doesn't look safe to me, either.

Reply:
To be safe, please copy shape values into `std::vector<uint64_t>` as use it from there. You can do it in one line with `std::vector(proto_tensor.dims().cbegin(), proto_tensor.dims().cend())`